---
title: "R Notebook"
output: html_notebook
---

# Population Structure  

Goal: to determine pop structure from VCF files. 

###Input files:  
   - We assume that all genotypes are stored as a VCF (see variant call format specifications: https://samtools.github.io/hts-specs/VCFv4.2.pdf), or BCF (binary VCF) file.  
   - There is no need to pre-qc these genotypes. This workflow will filter on minor allele frequency and call frequency. If you also want to filter on Hardy-Weinberg equilibrium, feel free to adjust as required.  
   
###Output:   
  - There are 2 key output files for this workflow:  
    - **population_structure.eigenvec**: contains the principal components, i.e. the genotypes after rotation. These can be plotted (e.g. PC1 vs. PC2) to visualise population-levels trends in the dataset  
    - **population_structure.eigenval**: contains the eigenvalues from the principal components analysis, i.e. the amount of variation explained by each principal component. 

###Population Structure Workflow:
  - filter the genotypes, removing pairs of SNPs which are in high linkage disequilibrium (LD) with each other.  
  - QC the genotypes & perform principal components analysis via plink  
  - produce scree plot of eigenvalues  
  - optionally, produce plots of the population structure
  
  
## Setup  

The workflow assumes that you are using plink2 (aka plink 1.9, available here: https://www.cog-genomics.org/plink2). This has been tested on Ubuntu 16.04. It will require tweaking to run on Windows machines.  

Please review the setup chunk below and make sure to set the root.dir option. This directory will become the working directory where all intermediates outputs are stored, so think about a sensible place for this (i.e. prob some scratch directory rather than a Git repo).
  
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/mnt/DataDrive/MedicagoFiles/Analyses/")
```

Below:  

  - declare you plink directory  
  - specify the directory where your genotype files are stored  
  - specify the working directory (all intermediate and final outputs will be placed here)  
  - review the workflow parameters, and change if necessary  
  - then, run the cell (will setup parameters, test plink installation and load necessary R packgaes)  
  
NOTE: that I have set the working_directory variable the same as the root.dir option in the setup chunk.  

```{r}
plink_ <- "/mnt/DataDrive/MedicagoFiles/Analyses/Pipeline/PLINK/plink"
genotype_directory <- "/mnt/DataDrive/MedicagoFiles/Genotypes/PublicDownload/"
working_directory <- "/mnt/DataDrive/MedicagoFiles/Analyses/PopulationStructure/"

# Genotype QC settings
parameters <- list(
    call_freq = 0.10,                       # rejects SNPs with more than this number of missing calls
    maf = 0.05,                             # rejects SNPs with a MAF < this threshold
    pop_struct_dimensions = 10,             # the number of principal components for population structure
    file_format = "bcf"
)


# test plink installation
if (!capture.output({ cat(system(sprintf("%s --version", plink_))) }) == 0) {

    cat(sprintf("
        --------------------------------------------------------------------------------

                                      !!!! WARNING !!!! 

        PLINK not found. Please review the plink_ directory in the Setup section.

        --------------------------------------------------------------------------------
    ", plink_))

}

# install & load packages
options(repos = "http://cran.uk.r-project.org")
if (!require(data.table)) {    
    install.packages("data.table")
}
if (!require(ggplot2)) {
    install.packages("ggplot2")
}

library(ggplot2)
library(data.table)
```

## Genotype QC and data conversion   

Plink can accept VCF and BCF files directly as inputs. However, in our case the Medicago BCF files do not have SNP identifiers (annoying - yes!), necessary for the LD pruning step. Below, we will perform the genotype QC and output plink-formatted (bed, bim and fam) files. For more information about plink formats please refer to:  https://www.cog-genomics.org/plink2/formats#bed  

The plink command is annotated below for reference:
```
plink --bcf <genotype_file> \\                     # points to a BCF file
        --const-fid \\                             # the Medicago data has no family structure, so we ste FID = 0 for all (this is a plink-ism)  
          --allow-extra-chr \\                     # fudge for the Medicago data
          --geno 0.1 \\                            # call frequency threshold
          --maf 0.05 \\                            # MAF threshold
          --make-bed \\                            # create plink-formatted bed, bim, fam files
          --set-missing-var-ids @:#[b37] \\        # create unique SNP ids (refer:https://www.cog-genomics.org/plink2/data#set_missing_var_ids)
          --out <output_file>                      # the name of the output file
```

Run the cell below. No changes necessary.  

This cell will read in all files in the gentoype_directory, which end in the specified file format (in our Medicago example, the filenames must end in "bcf"). For each of these files, it will QC the genotypes and write out plink formatted files to ```working_directory/plink_genotype_files/```.  

Review the cell below and run (no changes necessary).  

```{r}
genotype_files <- list.files(genotype_directory, pattern = sprintf("%s$", parameters$file_format), full.names = TRUE)
output_directory <- sprintf("%splink_genotype_files/", working_directory)

if (!dir.exists(output_directory)) {
    dir.create(file.path(output_directory))
}
for (genotype in genotype_files) {
    
    fileparts <- strsplit(genotype, "/")[[1]]
    filename <- strsplit(fileparts[length(fileparts)], "\\.")[[1]][1]
    
    cmd <- sprintf("
        %s --%s %s \\
           --const-fid \\
           --allow-extra-chr \\
           --geno %s \\
           --maf %s \\
           --make-bed \\
           --set-missing-var-ids @:#[b37] \\
           --out %s
    
    ", plink_, parameters$file_format, genotype, parameters$call_freq, parameters$maf,
       sprintf("%s%s", output_directory, filename))
    system(cmd)
    
    
}
```

The outputted plink files will have substantially less SNPs than the raw genotype files. The exact number of SNPs will be a reflection of the quality of the genotype calls.  


## LD Pruning  

For each plink genotype file, we will further reduce the number of SNPs based on pairwise LD. Li *et al* (2012, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2995003/) provide quite a nice overview of this process. In particular, the methods section states:  

"To reduce the LD between SNPs, PLINK was used to conduct SNP-pruning: we considered a sliding window of 1500 SNPs and calculated r2 between each pair of SNPs in that window; one of a pair of SNPs was removed if the pairwise r2 was greater than 0.2, then the window shifted 100 SNPs forward and the procedure was repeated. We used three-dimensional scatter plots of the top 3 significant principal components inferred by SMARTPCA to visualize population structure." (Li *et al*, 2012).

We are not going to be quite as stringent as this. We will consider a sliding window of 10 kilobases, with a slide of 10 kilobases and R2 threshold of 0.5.  

The following tutorial (http://www.jurgott.org/suzanne2012-07-23/GenABELexercise.pdf) (Wang & Leal, 2012) also has some nice notes with regards to determining population structure. In particular, the authors note that "for PCA and MDS you do not want to use markers which are in LD with each other. Here we only use a subset of GWAS markers for analysis of an actual GWAS data set would want to select a subset of markers which are not in LD" (p.8, Wang & Leal, 2012). This makes good sense, if you consider principal components analysis as a partitioning of covariance.  

Review the cell below and run (no changes necessary).

```{r}
genotype_directory <- output_directory
genotype_files <- unlist(lapply(list.files(genotype_directory, pattern = ".bed$", full.names = FALSE),
                                function (x) strsplit(x, "\\.")[[1]][1]))

for (genotype in genotype_files) {
    cmd <- sprintf("
               
        %s --bfile %s \\
           --indep-pairwise 10 10 0.5 \\
           --out %s
    ", plink_, sprintf("%s%s", genotype_directory, genotype), sprintf("%s%s_pruned_snps", working_directory, genotype))
    
    system(cmd)   
}
```

Again, this will result in a substantially reduced set of SNPs. Below, we will filter each genotype file and concatenate the results into one merged plink file.  

Review the chunk below, and then run (no changes necessary). 
**IMPORTANT** Review the snp_map output, and make sure that the genotypes and prune.in files map correctly - otherwise you will get nonsense later!

```{r}
genotype_directory <- output_directory
genotype_files <- list.files(genotype_directory, pattern = ".bed$", full.names = TRUE)
snp_files <- list.files(working_directory, pattern = "prune.in", full.names = TRUE)

# I am going assume (and test this to be confident!), that the genotype files will be in the 
# same order as the prune.in files. This seems like a safe assumption, since they will be alphabetical.
# HOWEVER... this should be reviewed!!!!

snp_map <- list()
for (idx in 1:length(snp_files)) {
    
    fileparts <- strsplit(genotype_files[[idx]], "/")[[1]]
    filename <- strsplit(fileparts[length(fileparts)], "\\.")[[1]][1]
    snp_map[[filename]] <- snp_files[idx]
}

print(snp_map)
```

If the ```snp_map``` is correct above, review the cell below and run.

```{r}

for (genotype in names(snp_map)) {
    
    tmp <- sprintf("%s%s",output_directory, genotype)
    
    idx <- which(names(snp_map) == genotype)
    fileparts <- strsplit(genotype_files[[idx]], "/")[[1]]
    filename <- strsplit(fileparts[length(fileparts)], "\\.")[[1]][1]
    
    filter_cmd <- sprintf("
    
        %s --bfile %s \\
           --extract %s \\
           --make-bed \\
           --out %s
    ", plink_, 
       sprintf("%s%s", genotype_directory, filename), 
       snp_map[[genotype]],
       sprintf("%s%s_ld_filtered", output_directory, filename))
    
    system(filter_cmd)
    
}

```

Finally, we will merge the filtered sets. Review the cell below and run (no changes required).
(this is ugly code - sorry.)

```{r}
# to merge plink files, we need to create a file listing the files we want to merge
merge_list <- file(sprintf("%sfiltered_files.txt", working_directory))
filtered_files <- list.files(output_directory, pattern = "ld_filtered.bed", full.names = TRUE)
filtered_files <- unlist(lapply(filtered_files, function (x) strsplit(x, "\\.")[[1]][1]))

write(filtered_files[-1], merge_list)
first_file <- strsplit(filtered_files[1], "\\.")[[1]][1]

# now merge them...    
combine_cmd <- sprintf("
    
    %s --bfile %s \\
       --merge-list %s  \\
       --make-bed \\
       --out %s
", plink_,
   first_file,
   sprintf("%sfiltered_files.txt", working_directory),
   sprintf("%smerged_genotypes", output_directory)
)

system(combine_cmd)

```

## Population structure  

Now we get to the good parts. Commonly, principal components analysis (PCA) or multidimensional scaling (MDS) are used to determine population structure. MDS requires us to first calculate genetic distances (via identity-by-descent (IBD)). IBD calculations take forever. PCA however, can be calculated relatively quickly directly from the genotypes. Everything I have read suggests that PCA and MDS can be used interchangeably for approximating population structure. So to keep things efficient, we will use PCA.  

Review the cell below and run (no changes required)
```{r}
cmd <- sprintf("
               
    %s --bfile %s  \\
       --allow-extra-chr \\
       --pca %s  \\
       --out %s
", plink_, 
   sprintf("%smerged_genotypes", output_directory), 
   parameters$pop_struct_dimensions,
   sprintf("%spopulation_structure", output_directory)
)

system(cmd)
```

This should produce 2 output files, ```population_structure.eigenval``` and ```population_structure.eigenvec``` (note, there will also be a log file). The eigenval file contains the eigenvalues for each of the prinicipal components, and describes the amount of variation attritbuted to each component. The eigenvec file contains the tranformed gentoypes (i.e. the principal components). As far as I know, plink does not output the rotation matrix.  

## Scree plot  

Plot of the eigenvalues against the factor (component) number.

```{r}
working_dir <- "/mnt/DataDrive/MedicagoFiles/Analyses/"
setwd(working_dir)

var_values <- read.csv("Medicago_Pop_Structure.eigenval", header = F)[, 1]
plot(var_values, xlab = "Factor", ylab = "Eigenvalue")
lines(var_values)
```



Looking at this, there is an elbow around 5 components. We'd probably get away with using this many quite nicely.

## Metaplot  

Here, we will plot the first two principal components and colour by various attributes that we know about the data, specifically: country of origin and flowering date  

```{r}
working_dir <- "/mnt/DataDrive/MedicagoFiles/Analyses/"
setwd(working_dir)
library(data.table)
library(ggplot2)

pca <- fread("Medicago_Pop_Structure.eigenvec")
meta <- fread("Pipeline/Traits/MedicagoTraits.combined")
country <- fread("../AccessionCountries.csv")

vis_data <- merge(pca[, 1:4, with = F], meta, by.x = c("V1", "V2"), by.y = c("FID", "IID"), all.x = TRUE)
vis_data <- merge(vis_data, country, by.x = "V2", by.y = "Accession", all.x = TRUE)
head(vis_data)


png("~/Documents/GitHub/GWASPipeline/Sandpit/PopulationStructure/Medicago_Population_Structure.png", height = 1000, width = 1000)
ggplot(vis_data, aes(x = V3, y = V4)) +
    geom_point(aes(colour = Country), alpha = 0.5, size = 5) +
    geom_text(aes(label = Country), colour = "darkblue", alpha = 0.6) +
    guides(colour = "none") +
    theme_minimal() +
    theme(axis.text = element_blank()) + xlab("") + ylab("") + ggtitle("Population Structure")
dev.off()
```


more